src/map.cpp

```diff
@@ -6992,3 +6992,138 @@ void map::creature_on_trap( Creature &c, bool const may_avoid )
     }
     tr.trigger( c.pos3(), &c );
 }
+
+template<typename Functor>
+    void map::function_over( const tripoint &start, const tripoint &end, Functor fun ) const
+{
+    function_over( start.x, start.y, start.z, end.x, end.y, end.z, fun );
+}
+
+template<typename Functor>
+    void map::function_over( const int stx, const int sty, const int stz, 
+                             const int enx, const int eny, const int enz, Functor fun ) const
+{
+    // start and end are just two points, end can be "before" start
+    // Also clip the area to map area
+    const int minx = std::max( std::min(stx, enx ), 0 );
+    const int miny = std::max( std::min(sty, eny ), 0 );
+    const int minz = std::max( std::min(stz, enz ), -OVERMAP_DEPTH );
+    const int maxx = std::min( std::max(stx, enx ), my_MAPSIZE * SEEX );
+    const int maxy = std::min( std::max(sty, eny ), my_MAPSIZE * SEEY );
+    const int maxz = std::min( std::max(stz, enz ), OVERMAP_HEIGHT );
+
+    // Submaps that contain the bounding points
+    const int min_smx = minx / SEEX;
+    const int min_smy = miny / SEEY;
+    const int max_smx = ( maxx + SEEX - 1 ) / SEEX;
+    const int max_smy = ( maxy + SEEY - 1 ) / SEEY;
+    // Z outermost, because submaps are flat
+    tripoint gp;
+    int &z = gp.z;
+    int &smx = gp.x;
+    int &smy = gp.y;
+    for( z = minz; z <= maxz; z++ ) {
+        for( smx = min_smx; smx <= max_smx; smx++ ) {
+            for( smy = min_smy; smy <= max_smy; smy++ ) {
+                submap const *cur_submap = get_submap_at_grid( smx, smy, z );
+                // Bounds on the submap coords
+                const int sm_minx = smx > min_smx ? 0 : minx % SEEX;
+                const int sm_miny = smy > min_smy ? 0 : miny % SEEY;
+                const int sm_maxx = smx < max_smx ? (SEEX - 1) : maxx % SEEX;
+                const int sm_maxy = smy < max_smy ? (SEEY - 1) : maxy % SEEY;
+
+                point lp;
+                int &sx = lp.x;
+                int &sy = lp.y;
+                for( sx = sm_minx; sx <= sm_maxx; ++sx ) {
+                    for( sy = sm_miny; sy <= sm_maxy; ++sy ) {
+                        const iteration_state rval = fun( gp, cur_submap, lp );
+                        if( rval != ITER_CONTINUE ) {
+                            switch( rval ) {
+                            case ITER_SKIP_ZLEVEL:
+                                smx = my_MAPSIZE + 1;
+                                smy = my_MAPSIZE + 1;
+                                // Fall through
+                            case ITER_SKIP_SUBMAP:
+                                sx = SEEX;
+                                sy = SEEY;
+                                break;
+                            default:
+                                return;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+void map::scent_blockers( bool blocks_scent[SEEX * MAPSIZE][SEEY * MAPSIZE],
```

Comment:
I think you want '''bool (&blocks_scent)[SEEX * MAPSIZE][SEEY * MAPSIZE]''', otherwise it degrades to a pointer with no size information.